hermes/apps/athena/modules/staked-ada-indexer/src/index.rs

```diff
@@ -0,0 +1,175 @@
+        pallas_addresses::{Address, ShelleyDelegationPart},
+        pallas_traverse::{MultiEraOutput, MultiEraTx},
+        MultiEraBlock, Network, StakeAddress,
+    },
+    database::staked_ada::{TxiByTxnIdRow, TxoAssetsByStakeRow, TxoByStakeRow},
+    utils::log::{error, warn},
+};
+
+/// Temporary buffers, holding entries to be inserted.
+#[derive(Default)]
+pub struct Buffers {
+    pub txo_by_stake: Vec<TxoByStakeRow>,
+    pub txo_assets_by_stake: Vec<TxoAssetsByStakeRow>,
+    pub txi_by_txn_id: Vec<TxiByTxnIdRow>,
+}
+
+impl Buffers {
+    /// Index all transaction inputs and outputs in a block.
+    pub fn index_block(
+        &mut self,
+        block: &MultiEraBlock,
+    ) {
+        for (txn_index, txn) in block.enumerate_txs() {
+            let txn_id = TransactionId::from(Blake2b256Hash::from(txn.hash()));
+            self.index_txn_txo(
+                block.network().clone(),
+                &txn,
+                block.slot().into(),
+                txn_id,
+                txn_index.into(),
+            );
+            self.index_txn_txi(&txn, block.slot().into());
+        }
+    }
+
+    /// Index the transaction outputs.
+    pub fn index_txn_txo(
+        &mut self,
+        network: Network,
+        txn: &MultiEraTx<'_>,
+        slot_no: u64,
+        txn_id: TransactionId,
+        txn_index: u16,
+    ) {
+        // Accumulate all the data we want to insert from this transaction here.
+        for (txo, txo_index) in txn.outputs().iter().zip(0u16..) {
+            // This will only return None if the TXO is not to be indexed (Byron Addresses).
+            // Skipping missing stake addresses: only indexing staked ADA and assets.
+            let Some((Some(stake_address), _)) =
+                extract_stake_address(network.clone(), txo, slot_no)
+            else {
+                continue;
+            };
+
+            self.txo_by_stake.push(TxoByStakeRow {
+                stake_address: stake_address.clone(),
+                txn_id,
+                txn_index,
+                txo: txo_index,
+                slot_no,
+                value: txo.value().coin().into(),
+                spent_slot: None,
+            });
+
+            for asset in txo.value().assets() {
+                let policy_id = *asset.policy();
+                for policy_asset in asset.assets() {
+                    if policy_asset.is_output() {
+                        self.txo_assets_by_stake.push(TxoAssetsByStakeRow {
+                            stake_address: stake_address.clone(),
+                            slot_no,
+                            txn_index,
+                            txo: txo_index,
+                            policy_id,
+                            asset_name: policy_asset.name().to_vec(),
+                            value: policy_asset.any_coin().into(),
+                        });
+                    } else {
+                        error!("Minting MultiAsset in TXO.");
```

Comment:
Note to self, create an issue to de-duplicate the code in `Athena` and `Gateway`.